util/common/src/main/java/org/hibernate/search/util/common/data/Range.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.util.common.data;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+
+import org.hibernate.search.util.common.impl.Contracts;
+
+/**
+ * A representation of a range that can be used with any type.
+ * <p>
+ * Because there are no restrictions on type of values that can be used with this {@link Range} class,
+ * it is not able to "understand" values that are passed to its various factory methods.
+ * As a result, only minimal consistency checks are performed: null-checks, mostly.
+ * In particular, <strong>this class does not check that the lower bound is actually lower than the upper bound</strong>,
+ * because it has no idea what ordering to use.
+ * Checking the relative order of bounds is the responsibility of callers of the {@link #getLowerBoundValue()}
+ * and {@link #getUpperBoundValue()} methods.
+ *
+ * @param <T> The type of values in this range.
+ */
+public final class Range<T> {
+
+	/**
+	 * @param lowerBoundValue The lower bound of the range.
+	 * May be {@code null} to represent {@code -Infinity} (no lower bound),
+	 * @param upperBoundValue The upper bound of the range.
+	 * May be {@code null} to represent {@code +Infinity} (no upper bound).
+	 * @param <T> The type of range bounds.
+	 * @return The range {@code [lowerBoundValue, upperBoundValue)} (lower bound included, upper bound excluded),
+	 * or {@code (-Infinity, upperBoundValue)} (both bounds excluded) if the lower bound is {@code -Infinity}.
+	 */
+	public static <T> Range<T> of(T lowerBoundValue, T upperBoundValue) {
+		return new Range<>( lowerBoundValue, RangeBoundInclusion.INCLUDED, upperBoundValue, RangeBoundInclusion.EXCLUDED );
```

Comment:
It wouldn't be simpler for us, because Elasticsearch range aggregations include the lower bound and exclude the upper bound.
But more importantly, I'm not sure it would be simpler for users. Generally user need disjoint, adjacent ranges in their range aggregations:

* either `[0, 10)`, `[10, 20)`, etc.
* or  `(0, 10]`, `(10, 20]`, etc.

Including or excluding both bounds by default would result in impractical ranges:

* if included by default, there are overlaps: `[0, 10]`, `[10, 20]`, etc.
* if excluded by defaults, there are gaps:  `(0, 10)`, `(10, 20)`, etc.

Does this make sense?

Reply:
It does. thanks